[DAG] Support saturated truncate

ParkHanbum opened the pull request:
A truncate is considered saturated if no additional conversion is
required between the target and return values. If the target is
saturated when attempting to truncate from a vector, there is an
opportunity to optimize it.

Previously, each architecture had its own attempt at optimization,
leading to redundant code. This patch implements common logic by
introducing three new ISDs:

 `ISD::TRUNCATE_SSAT_S`: When the operand is a signed value and
 the range of values matches the range of signed values of the
 destination type.

 `ISD::TRUNCATE_SSAT_U`: When the operand is a signed value and
 the range of values matches the range of unsigned values of the
 destination type.

 `ISD::TRUNCATE_USAT_U`: When the operand is an unsigned value and
 the range of values matches the range of unsigned values of the
 destination type.

These ISDs indicate a saturated truncate.

Fixes https://github.com/llvm/llvm-project/issues/85903

ParkHanbum committed:
[AArch64] Support saturated truncate

Add support for saturated truncate with the following changes:

- Add action to Legal for types v8i16, v4i32, and v2i64
- Implement `isTypeDesirableForOp` to check for truncate conversions
- Add patterns for saturated truncate of supported types

ParkHanbum committed:
[RISCV] Support saturated truncate

Add support for saturated truncate by implementing the following changes:

- Add `TRUNCATE_[SU]SAT_[SU]` to the Action target of `TRUNCATE`
- Add `TRUNCATE_[SU]SAT_[SU]` to the TargetLowering target of `TRUNCATE`
- Convert `TRUNCATE_SSAT_S` to `TRUNCATE_VECTOR_VL_SSAT`
- Convert `TRUNCATE_[SU]SAT_U` to `TRUNCATE_VECTOR_VL_USAT`

topperc commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -804,6 +804,9 @@ enum NodeType {
   /// TRUNCATE - Completely drop the high bits.
   TRUNCATE,
+  /// TRUNCATE_[SU]SAT - Truncate for saturated operand
+  TRUNCATE_SSAT,
```
Is TRUNCATE_SSAT signed input to signed result?
Is TRUNCATE_USAT unsigned input to unsigned result?

I ask because X86's packuswb instructon does signed input to unsigned result so its worth being clearing here.

ParkHanbum replied:
*This is my opinion and needs confirmation*. 
It is true that SSAT/USAT is signed, but 'truncate_[us]sat' means that the range of values for 'truncate' is in the range of values that don't have to care about the sign bit.

We also don't support unsigned type variables, right? Therefore, the 'truncate_sat' instruction doesn't care about the type of target value, just convert it to return type and return it.

topperc replied:
Variables don't have signed or unsigned type but operations can interpret their input as signed or unsigned.

I believe as you have defined it TRUNCATE_USAT will interpret the input as an unsigned value and produce an unsigned result in the destination type.

The x86 packuswb interprets the input as signed 16 bits and produces an unsigned 8 bit result. If the input is negative it will return 0. That is a different operation than either operation defined here. It's equivalent to smax with 0 followed by truncate_usat. I could imagine having that operation as a single node. Not suggesting for this patch.

I just want the semantics of the new opcodes documented.

davemgreen replied:
AArch64 has a SQXTUN node too: https://docsmirror.github.io/A64/2023-09/sqxtun_advsimd.html

RKSimon replied:
So should we have 3 truncsat nodes? I mentioned this on #85903 but hadn't realised other targets had something similar to PACKUS
```
TRUNCATE_SSAT_S,  // saturate signed input to signed result - truncate(smax(smin(x)))
TRUNCATE_SSAT_U,  // saturate signed input to unsigned result - truncate(umax(smin(x,0)))
TRUNCATE_USAT_U,  // saturate unsigned input to unsigned result - truncate(umin(x))
```

ParkHanbum replied:
@RKSimon should add that ISD? is it not necessary for unsigned input to signed input?

RKSimon replied:
If we have targets that support for TRUNCATE_USAT_S then by all means support it.

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14867,13 +14872,138 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  return SDValue();
+}
+
+SDValue DAGCombiner::visitTRUNCATE_SSAT(SDNode *N) { return SDValue(); }
```
Can we just not add the visit function instead of adding an empty function?

topperc commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -5343,9 +5343,13 @@ def VImm8000: PatLeaf<(AArch64mvni_msl (i32 127), (i32 264))>;
 // trunc(umin(X, 255)) -> UQXTRN v8i8
 def : Pat<(v8i8 (trunc (umin (v8i16 V128:$Vn), (v8i16 VImmFF)))),
           (UQXTNv8i8 V128:$Vn)>;
+def : Pat<(v8i8 (truncusat (v8i16 V128:$Vn))),
```
Can the pattern above be dropped?

ParkHanbum replied:
This is not possible with the current logic. However, I think it would be possible with additional logic. Would you like to see it included in this patch?

topperc replied:
What are the t18 and t21 nodes. It looks like you only pasted part of the DAG.

ParkHanbum replied:
@topperc That's weird, but I didn't touch anything else.

ParkHanbum replied:
@topperc I tested this by implementing additional code to support `fp_to_[su]int_sat`. As a result, if `LegalOperations` is unchecked, the associated Pattern can be deleted. 

Should we continue to check `LegalOperations`? I can not sure because I'm not a veteran.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -16617,9 +16631,11 @@ static SDValue combineTruncToVnclip(SDNode *N, SelectionDAG &DAG,
-  if ((Val = DetectUSatPattern(Src)))
+
+  Val = N->getOperand(0);
+  if (N->getOpcode() == RISCVISD::TRUNCATE_VECTOR_VL_USAT)
```
Is TRUNCATE_VECTOR_VL_USAT the same as RISCVISD::VNCLIPU_VL?

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -16617,9 +16631,11 @@ static SDValue combineTruncToVnclip(SDNode *N, SelectionDAG &DAG,
+  if (N->getOpcode() == RISCVISD::TRUNCATE_VECTOR_VL_USAT)
     ClipOpc = RISCVISD::VNCLIPU_VL;
-  else if ((Val = DetectSSatPattern(Src)))
+  else if (N->getOpcode() == RISCVISD::TRUNCATE_VECTOR_VL_SSAT)
```
Is TRUNCATE_VECTOR_VL_SSAT the same as VNCLIP_VL?

topperc replied:
Nevermind. It's not the same because VNCLIP has a shift amount.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -5375,6 +5383,10 @@ def : Pat<(v16i8 (concat_vectors
                  (v8i8 (trunc (smax (smin (v8i16 V128:$Vn), (v8i16 VImm7F)),
                                           (v8i16 VImm80)))))),
           (SQXTNv16i8 (INSERT_SUBREG (IMPLICIT_DEF), V64:$Vd, dsub), V128:$Vn)>;
+def : Pat<(v16i8 (concat_vectors
```
It looks like we are missing some patterns for the unsigned upper "2" cases, I will see about adding them and they can be witched over too. Hopefully the new node will be able to handle i64->i32 cases too.

davemgreen replied:
There are some extra test cases in llvm/test/CodeGen/AArch64/qmovn.ll (but only for legal cases), and I added the missing patterns in cbebacef5e12c985ca32c894bcfddec9adb2fc85.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1274,6 +1274,8 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::AVGCEILU, VT, Legal);
       setOperationAction(ISD::ABDS, VT, Legal);
       setOperationAction(ISD::ABDU, VT, Legal);
+      setOperationAction(ISD::TRUNCATE_SSAT, VT, Legal);
```
Is the intent to say that a `MVT::v8i16 -> MVT::v8i8` truncating saturate is legal? If so does it need to mark the `MVT::v8i16` or the `MVT::v8i8`, and what happens if it is a `MVT::v8i32 -> MVT::v8i8` or `MVT::v4i32 -> MVT::v4i8` truncate?

If the isOperationLegalOrCustom looks at SrcVT then we likely only need to mark `MVT::v8i16, MVT::v4i32, MVT::v2i64` as legal, but there should hopefully be tests for the other case mentioned above. Note that I don't think that v2i64->v2i32 is handled at the moment with our tablegen patterns because we don't have support for 64bit min/max. So hopefully they might be able to start working with this patch.

ParkHanbum replied:
@davemgreen got it. I will change it to custom so that it is handled properly when legalizing the operand.

topperc commented:
I just submitted #100173 to remove the RISCVISD::VNCLIP* opcodes in favor of the TRUNCATE_VECTOR_VL_*SAT nodes added in this patch. This should allow us to remove the translating from TRUNCATE_VECTOR_VL_*SAT to RISCVISD::VNCLIP*.

davemgreen commented on llvm/test/CodeGen/AArch64/qmovn.ll:
```diff
@@ -292,15 +292,15 @@ entry:
 ; Test the (concat_vectors (X), (trunc(umin(smax(Y, 0), 2^n))))) pattern.
 
+; TODO: %min is a value between 0 and 255 and is within the unsigned range of i8.
+; So it is saturated truncate. we have an optimization opportunity.
```
I'm pretty sure this should be matching a sqxtun2, that was the intent. I believe because the lower limit is already clamped, that the upper smin is equivalent to umin.
https://godbolt.org/z/e7ne31TYb
You can see in that example that the midend turned smin into umin.

ParkHanbum replied:
(right place :))
agree. I also added a comment with that intention and I'm thinking maybe adding code to DAGCombiner can solve it.
do you think I should include this in this patch as well?

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1410,6 +1410,12 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
   }
 
+  for (MVT VT : {MVT::v8i16, MVT::v4i32}) {
+    setOperationAction(ISD::TRUNCATE_SSAT_S, VT, Custom);
```
If this uses isTypeDesirableForOp to only generate the truncates of the expected size, then these can be legal as opposed to custom. (If they were custom then there would need to be something in LowerOperation to handle them).

ParkHanbum replied:
agree. I also added a comment with that intention and I'm thinking maybe adding code to `DAGCombiner` can solve it. 

do you think I should include this in this patch as well?

davemgreen replied:
Sorry - I'm not sure I understand what you mean. I believe in this version of the code just making them Legal should be fine, but I'm not sure which comment this was referring to or what code to DAG combiner.

ParkHanbum replied:
never mind. I was left comment at wrong place

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,159 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+SDValue DAGCombiner::visitTRUNCATE_USAT(SDNode *N) {
+  EVT VT = N->getValueType(0);
+  SDValue N0 = N->getOperand(0);
+  SDValue FPInstr = N0.getOpcode() == ISD::SMAX ? N0.getOperand(0) : N0;
```
Do you need to check the other operand of this SMAX?

ParkHanbum replied:
Yes, missed it. I'll fix it.

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,159 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+    SDValue Sat = DAG.getNode(ISD::FP_TO_UINT_SAT, SDLoc(FPInstr), VT,
+                              FPInstr.getOperand(0),
+                              DAG.getValueType(VT.getScalarType()));
+    return Sat;
```
return DAG.getNode... no need for temporary variable.

ParkHanbum replied:
I'm not sure how I can do it. could you give me little advise please?

ParkHanbum replied:
@topperc how about now? do you think I'm doing properly?

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14930,6 +15088,11 @@ SDValue DAGCombiner::visitTRUNCATE(SDNode *N) {
+  // fold satruated truncate
+  if (SDValue SaturatedTR = foldToSaturated(N, VT, N0, SrcVT, DL, TLI, DAG)) {
+    return SaturatedTR;
+  }
```
Drop curly braces

ParkHanbum commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1410,6 +1410,12 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     }
   }
 
+  for (MVT VT : {MVT::v8i16, MVT::v4i32}) {
```
I can add it, but I don't know which pattern is suitable for it.

davemgreen replied:
I think the basics should be this, with the different sat nodes.
```
def : Pat<(v2i32 (truncusat_u (v2i64 V128:$Vn))),
          (UQXTNv2i32 V128:$Vn)>;
```

(There are other "concat" UQXTN2 patterns but the ones above are the basics needed to get things working).

ParkHanbum replied:
@davemgreen can you check it?  v2i64 -> v2i32 is correct?

tschuett commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -814,6 +814,13 @@ enum NodeType {
   TRUNCATE,
+  /// TRUNCATE_[SU]SAT - Truncate for saturated operand
+  TRUNCATE_SSAT_S, // saturate signed input to signed result -
+                   // truncate(smin(smax(x)))
```
Could you please spell out the parameters for my eyes? [s|u][min|max] have two parameters.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -5418,64 +5418,60 @@ def VImm7FFF: PatLeaf<(AArch64movi_msl (i32 127), (i32 264))>;
+def : Pat<(v4i16 (truncusat_u (v4i32 V128:$Vn))),
           (UQXTNv4i16 V128:$Vn)>;
 // trunc(smin(smax(X, -128), 128)) -> SQXTRN
 //  with reversed min/max
```
The "reversed min/max" lines can be removed now.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,159 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+         "Unexpected types for truncate operation");
+
+  // Match min/max and return limit value as a parameter.
+  auto MatchMinMax = [](SDValue V, unsigned Opcode, APInt &Limit) -> SDValue {
```
MatchMinMax can be made a separate function so that it is shared between all the three new routines, if they work the same way (either return the constant and check it later or check for the correct constant).

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,159 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  APInt C1, C2;
+  if (SDValue SMin = MatchMinMax(In, ISD::SMIN, C2))
+    if (MatchMinMax(SMin, ISD::SMAX, C1))
+      if (C1.isNonNegative() && C2.isMask(VT.getScalarSizeInBits()))
```
I would expect C1==0, to handle the basic case. Then it can check C1 >=s 0 and C2 isMask, and return the "SMax" value like the others.

Then - because the SMAX is making sure the value is non-negative we can match against either SMIN or UMIN, they should both be valid.

It looks like this is attempting to handle more cases - they might be good to add in a follow-up once the basics are in.

davemgreen reviewed (commented):
Yeah the improvements look good, it's good to see it handles more types now. If we can get the TRUNCATE_SSAT_U nodes working then this sounds OK. It should match from either `smax(smin(x, 255), 0)` or `smax(umin(x, 255), 0)`

ParkHanbum commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -5418,64 +5418,62 @@ def VImm7FFF: PatLeaf<(AArch64movi_msl (i32 127), (i32 264))>;
-                              (v4i32 VImm8000)))),
+// trunc(smin(smax(X, -2147483648), 2147483647)) -> SQXTRN
+def : Pat<(v2i32 (truncssat_s (v2i64 V128:$Vn))),
           (SQXTNv4i16 V128:$Vn)>;
```
sure. should I write the tests related to it?

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,166 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  // where C1 >= 0 and C2 is unsigned max of destination type.
+  if (SDValue SMin = matchMinMax(In, ISD::SMIN, C2, /*Signed*/ false))
+    if (matchMinMax(SMin, ISD::SMAX, C1, /*Signed*/ false))
+      if (C1.isNonNegative() && C2.isMask(VT.getScalarSizeInBits()))
+        return SMin;
```
Should this just be:
```
  if (C1 == 0 && C2.isMask(VT.getScalarSizeInBits()))
     return SMax;
```

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,166 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  // where C1 >= 0, C2 is unsigned max of destination type and C1 <= C2.
+  if (SDValue SMax = matchMinMax(In, ISD::SMAX, C1, /*Signed*/ false))
+    if (SDValue SMin = matchMinMax(SMax, ISD::SMIN, C2, /*Signed*/ false))
+      if (C1.isNonNegative() && C2.isMask(VT.getScalarSizeInBits()) &&
+          C2.uge(C1))
+        return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
```
And similar here:
```
if (C1 == 0 && C2.isMask(VT.getScalarSizeInBits()))
  return SMin;
```

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,166 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+    if (C2.isMask(VT.getScalarSizeInBits())) {
+      // (truncate (umin (smax (x, C1), C2)))
+      // where C1 == 0, C2 is unsigned max of destination type.
+      if (SDValue SMax = matchMinMax(UMin, ISD::SMAX, C1, /*Signed*/ false)) {
```
I think this logic should be in detectSSatUPattern.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,181 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  else if (First = UMin = matchMinMax(In, ISD::UMIN, Max, /*Signed*/ false))
+    SMax = matchMinMax(UMin, ISD::SMAX, Min, /*Signed*/ false);
+
+  if (SMax && Min.isNonNegative() && Max.isMask(VT.getScalarSizeInBits())) {
```
Can you explain this Min.isNonNegative() bit, and why it is needed?

ParkHanbum replied:
This is because we want to check the range of unsigned values. The minimum value must be greater than or equal to zero.

davemgreen replied:
I feel that the non-negative part is complicating things. Can we keep this patch simpler and just match the exact pattern. Something like:
```
  APInt UnsignedMax, Zero;
  UnsignedMax = APInt::getMaxValue(NumDstBits).zext(NumSrcBits);
  Zero = APInt::getZero(NumSrcBits);

  // (truncate (smin (smax (x, 0), Max)) to dest_type),
  if (Max = matchMinMax(In, ISD::SMAX, Zero)) {
    if (Min = matchMinMax(Max, ISD::SMIN, UnsignedMax))
      return Min;
    if (Min = matchMinMax(Max, ISD::UMIN, UnsignedMax))
      return Min;
  }
  if (Min = matchMinMax(In, ISD::SMIN, UnsignedMax))
    if (Max = matchMinMax(Min, ISD::SMAX, Zero))
      return Max;
  if (Min = matchMinMax(In, ISD::UMIN, UnsignedMax))
    if (Max = matchMinMax(Min, ISD::SMAX, Zero))
      return Max;

  return SDValue();
```

We can add the extra bits in a follow-up if they are needed.

ParkHanbum replied:
how about now? do you think I'm doing properly?

ParkHanbum replied:
we need to considering when Min > 0. so, we need to get the Min value from matchMinMax.

ParkHanbum replied:
we need to considering when Min > 0. 
so, &C was not removed to get the Min value from matchMinMax.

ParkHanbum replied:
@davemgreen how about this? 
```
static SDValue matchMin(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C))
      if (C == Limit)
        return V.getOperand(0);
      if (C > Limit && Limit == 0)
        return V.getOperand(0);
  }
  return SDValue();
}
```

ParkHanbum replied:
```
  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax))
      return SMin.getOperand(0);
  }
  // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO)) {
    if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(In, ISD::SMAX, Zero))
      return SMax.getOperand(0);
    // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO))
      return SMin;
  }
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(In, ISD::SMAX, Zero))
      return SMax.getOperand(0);
    // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO))
      return UMin;
  }
```
I made a temporary code of what it would look like if I changed to the logic you showed in your example, where I want to modify the part that says `OVER_ZERO` to check if `Min>0`. What do you think?
@davemgreen

ParkHanbum replied:
```
// Match min and return limit value as a parameter.
static SDValue matchMinMax(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C)) {
      if (C == Limit)
        return V.getOperand(0);
      if (V.getOpcode() == ISD::SMAX && !Limit.isNegative() && C.sgt(Limit) && Limit != 0)
        return V.getOperand(0);
    }
  }
  return SDValue();
}
```
I've solved the above homework by excluding zero cases from `matchMinMax`, what do you think?

ParkHanbum replied:
```

  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return SMax.getOperand(0);
  }
  // Min >= 1, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin.getOperand(0);
    // Min >= 1, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin;
  }
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return UMin.getOperand(0);
    // Min >= 1, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return;
```
here is full code

ParkHanbum replied:
```

// Match min and return limit value as a parameter.
static SDValue matchMinMax(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C)) {
      if (C == Limit)
        return V.getOperand(0);
      if (V.getOpcode() == ISD::SMAX && !Limit.isNegative() && C.uge(Limit) &&
          Limit != 0)
        return V.getOperand(0);
    }
  }
  return SDValue();
}
```
```

  APInt One = APInt(NumSrcBits, 1);
  
  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return SMax.getOperand(0);
  }
  // Min >= 1, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, One)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin.getOperand(0);
    // Min >= 1
    if (SMax = matchMinMax(SMin, ISD::SMAX, One))
      return SMin;
  }
  // Max is unsigned max of destination type
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return UMin.getOperand(0);
    if (SMax = matchMinMax(UMin, ISD::SMAX, One))
      return UMin;
  }
  ```
I've solved the above homework by excluding zero cases from `matchMinMax`, what do you think?

davemgreen replied:
Can we just keep the basic version in this first patch? The patterns we want to match are:
USAT_U: `trunc(umin(x, 255))`
SSAT_S: `trunc(smin(smax(x, -128), 127))`
SSAT_U: `trunc(smin(smax(x, 0), 255))`

The min/max could be in the opposite orders, and the last one is odd because it can match `trunc(umin(smax(x, 0), 255))` as the umin and smin are equivalent.
I don't believe we should need any >= checks at the moment, and if we do they need tests. IMO they would be better adding in a follow-up patch, so we can get the basics down first. They might well enable the creating of more saturating nodes, but with this patch adding all three variants there is already quite a lot going on.

davemgreen reviewed (commented):
Thanks - the AArch64 test and patterns all look good to me now. I can't say much about the RISCV parts and I think there might be other (untested) cases where things can still go wrong with TRUNCATE_SSAT_U.

ParkHanbum committed:
[AArch64] Add fail tests with unsupported types in saturated truncate

ParkHanbum commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,181 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  } else
+    FPInstr = MatchFPTOINT(N0);
+
+  if (FPInstr) {
```
thanks @davemgreen always.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,180 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  EVT VT = N->getValueType(0);
+  SDValue N0 = N->getOperand(0);
+
+  std::function<SDValue(SDValue)> MatchFPTOINT = [&](SDValue Val) -> SDValue {
```
This looks like it would match `trunc_usat_u(fptosi)`, `trunc_usat_u(fptoui)`, `trunc_usat_s(fptosi)` and `trunc_usat_s(fptoui)`, converting them all to `fptoui_sat`. I think `trunc_usat_u(fptoui)` should be enough and the others might clamp to the wrong range.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,180 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+}
+
+/// Detect patterns of truncation with signed saturation:
+/// (truncate (smin ((smax (x, signed_min_of_dest_type)),
+///                  signed_max_of_dest_type)) to dest_type)
```
`((smax (x, signed_min_of_dest_type)),` -> `(smax (x, signed_min_of_dest_type),`

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,180 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+
+  APInt Max;
+  APInt UnsignedMax = APInt::getMaxValue(NumDstBits).zext(NumSrcBits);
+  if (SDValue UMin = matchMin(In, UnsignedMax, Max)) {
```
Could this now match from a SMIN? I think I liked the way matchMinMax worked in an earlier revision where it was given an Opcode to match.

ParkHanbum replied:
it matched well. but it doesn't matter, I'll recover earlier revision

ParkHanbum replied:
it matched. but if you want see opcode what we matched I'll recover it.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,171 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C)) {
+      if (C == Limit)
+        return V.getOperand(0);
+      if (V.getOpcode() == ISD::SMAX && !Limit.isNegative() && C.uge(Limit) &&
+          Limit != 0)
+        return V.getOperand(0);
```
Remove this SMAX special case.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14915,6 +14920,153 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+  assert(NumSrcBits > NumDstBits && "Unexpected types for truncate operation");
+
+  APInt UnsignedMax = APInt::getMaxValue(NumDstBits).zext(NumSrcBits);
+  if (SDValue Min = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
```
These could all be done more cleanly with sd_match:
```cpp
SDValue Min;
if (sd_match(In, m_UMin(m_Value(Min), m_SpecificInt(UnsignedMask))))
  return Min;
```

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14992,50 +14978,46 @@ static SDValue detectSSatSPattern(SDValue In, EVT VT) {
-      return Min.getOperand(0);
-  }
+  if (sd_match(In, m_SMax(m_SMin(m_Value(Val), m_SpecificInt(UnsignedMax)),
+                          m_SpecificInt(Zero))))
```
You can use m_Zero() instead of m_SpecificInt(Zero)

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -14930,6 +15061,10 @@ SDValue DAGCombiner::visitTRUNCATE(SDNode *N) {
   if (N0.getOpcode() == ISD::TRUNCATE)
     return DAG.getNode(ISD::TRUNCATE, DL, VT, N0.getOperand(0));
 
+  // fold satruated truncate
```
satruated -> saturated

RKSimon commented:
Please can you edit the PR summary to describe the 3 nodetypes, it still refers to the original 2

davemgreen reviewed (commented):
Sorry I wasn't around yesterday, thanks for the updates. The code looks good to me now, if others agree.

ParkHanbum commented:
@RKSimon @davemgreen Is it okay to reduce commits?

ParkHanbum committed:
[DAG] Support saturated truncate

A truncate is considered saturated if no additional conversion is
required between the target and return values. If the target is
saturated when attempting to truncate from a vector, there is an
opportunity to optimize it.

Previously, each architecture had its own attempt at optimization,
leading to redundant code. This patch implements common logic by
introducing three new ISDs:

 `ISD::TRUNCATE_SSAT_S`: When the operand is a signed value and
 the range of values matches the range of signed values of the
 destination type.

 `ISD::TRUNCATE_SSAT_U`: When the operand is a signed value and
 the range of values matches the range of unsigned values of the
 destination type.

 `ISD::TRUNCATE_USAT_U`: When the operand is an unsigned value and
 the range of values matches the range of unsigned values of the
 destination type.

These ISDs indicate a saturated truncate.

Fixes #85903

ParkHanbum committed:
[DAG] Combine `truncusat_u(fptoui(x))` to `fptoui_sat(x)`

ParkHanbum committed:
[AArch64] Add tests for saturated truncate

RKSimon commented:
> @RKSimon @davemgreen Is it okay to reduce commits?

Sure - as I mentioned above the commit summary at the top needs editing. Once that's done I'll approve.

davemgreen commented:
Can you update the description in github with the one in https://github.com/llvm/llvm-project/pull/99418/commits/8d81896a8b1f1137a2a182597551047c1092b1da (or the others if you want to combine them)? LLVM uses a squash-and-merge approach to committing PR's, so they will be squashed into a single commit.

ParkHanbum commented:
do you mean my written message when I request PR? if it is, I done.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -1908,6 +1909,8 @@ SDValue DAGCombiner::visit(SDNode *N) {
   case ISD::ANY_EXTEND_VECTOR_INREG: return visitEXTEND_VECTOR_INREG(N);
   case ISD::TRUNCATE:           return visitTRUNCATE(N);
+  case ISD::TRUNCATE_SSAT_U:
+  case ISD::TRUNCATE_USAT_U:    return visitTRUNCATE_USAT(N);
```
Should this be called visitTRUNCATE_SAT_U? Or should we just have a visitTRUNCATE_SAT call and handle the unsigned cases inside it?

davemgreen replied:
Did we not mean to change this to just `case ISD::TRUNCATE_USAT_U: return visitTRUNCATE_USAT(N);`, without the TRUNCATE_SSAT_U case? (Sorry if I missed that). Otherwise it will change `TRUNCATE_SSAT_U(FP_TO_UINT(x))` to `FP_TO_UINT_SAT(x)`, which will not clamp to the same bounds.

ParkHanbum replied:
visitTRUNCATE_USAT has a small task, so I don't see the need to separate SSAT and USAT.
Is it LLVM's way to reduce unnecessary separation and separate them if they are reasonably large?

RKSimon replied:
No, we can generalize it later if the need arises - but we need to confirm if we should be handling the TRUNCATE_SSAT_U  case or not (do we have test coverage?)

ParkHanbum replied:
If that's the case, then we don't have a test for truncate_ssat_u. As @davemgreen commented, it is right to change to call visitTRUNCATE_USAT() in case TRUNCATE_USAT_U.

ParkHanbum replied:
should I change it so that only ` case ISD::TRUNCATE_USAT_U: return visitTRUNCATE_USAT(N);` remains?

davemgreen replied:
I think so. Perhaps call it visitTRUNCATE_USAT_U too?

RKSimon reviewed (commented):
one final minor

RKSimon reviewed (approved):
LGTM

davemgreen reviewed (approved):
Thanks. LGTM

RKSimon commented:
@ParkHanbum Are you happy for me to commit this?

ParkHanbum commented:
@RKSimon I feel very honored, Sir!

RKSimon reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8279,11 +8287,18 @@ SDValue RISCVTargetLowering::lowerVectorTruncLike(SDValue Op,
+  unsigned NewOpc;
+  if (Opc == ISD::TRUNCATE_SSAT_S)
+    NewOpc = RISCVISD::TRUNCATE_VECTOR_VL_SSAT;
+  else if (Opc == ISD::TRUNCATE_SSAT_U || Opc == ISD::TRUNCATE_USAT_U)
```
How are we lowering 2 differnet opcodes to the same RISC-V instruction?
